recurly.go

```diff
@@ -0,0 +1,489 @@
+		}
+	}
+}
+
+// parses client errors.
+func (r *response) parseClientError(v interface{}) error {
+	// Immediately return a client error if there is no response body.
+	if r.Header.Get("Content-Length") == "0" {
+		return &ClientError{Response: r.Response}
+	}
+
+	// Read the full response body so we can conditionally process the
+	// xml based on the top level tag that is returned.
+	b, err := ioutil.ReadAll(r.Response.Body)
+	if err != nil {
+		return err
+	} else if len(b) == 0 {
+		// Exit here to avoid io.EOF errors.
+		return &ClientError{Response: r.Response}
+	}
+
+	var name struct {
+		XMLName xml.Name
+	}
+	if err := xml.Unmarshal(b, &name); err != nil {
+		return err
+	}
+
+	switch name.XMLName.Local {
+	case "error":
+		var e xmlSingleError
+		if err := xml.Unmarshal(b, &e); err != nil {
+			return err
+		}
+		return &ClientError{
+			Response: r.Response,
+			ValidationErrors: []ValidationError{{
+				Description: e.Description,
+				Field:       e.Field,
+				Symbol:      e.Symbol,
+			}},
+		}
+	case "errors":
+		var e xmlMultiErrors
+		if err := xml.Unmarshal(b, &e); err != nil {
+			return err
+		}
+
+		// Any transaction errors return TransactionFailedError.
+		if e.Transaction != nil || e.TransactionError != nil {
+			transFailedErr := &TransactionFailedError{
+				Response:    r.Response,
+				Transaction: e.Transaction,
+			}
+
+			if e.TransactionError != nil {
+				transFailedErr.TransactionError = *e.TransactionError
+			}
+			return transFailedErr
+		}
+
+		clientErr := &ClientError{Response: r.Response}
+		if e.Error.Description != "" || e.Error.Field != "" || e.Error.Symbol != "" {
+			clientErr.ValidationErrors = []ValidationError{{
+				Description: e.Error.Description,
+				Field:       e.Error.Field,
+				Symbol:      e.Error.Symbol,
+			}}
+		}
+		return clientErr
+	}
+
+	// Unknown body.
+	return &ClientError{Response: r.Response}
+}
+
+// Rate represents the rate limit for the current client.
+type Rate struct {
+	// The total request limit during the 5 minute window (e.g. requests/min * 5 min)
+	Limit int
+
+	// The number of requests remaining until your requests will be denied.
+	Remaining int
+
+	// The time when the current window will completely reset assuming no further API requests are made.
+	Reset time.Time
+}
+
+// RateLimitError occurs when Recurly returns a 429 Too Many Requests error.
+type RateLimitError struct {
+	Response *http.Response
+
+	Rate Rate // Rate specifies the last known rate limit for the client
+}
+
+func (e *RateLimitError) Error() string {
+	return fmt.Sprintf("API rate limit exceeded: %s %s: %d %v",
+		e.Response.Request.Method,
+		e.Response.Request.URL.Path,
+		e.Response.StatusCode,
+		e.Rate.Reset.Sub(time.Now()),
+	)
+}
+
+// ClientError occurs when Recurly returns 400-499 status code.
+// There are two known exceptions to this:
+// 1) 429 Too Many Requests. See RateLimitError.
+// 2) 422 Unprocessable Entity with a failed transaction. See TransactionFailedError.
```

Comment:
this comment is not accurate.  there're cases recurly returns `422` status code but the error cause/response has nothing to do with transaction. here is an example i experienced:
```
Status Code: 422<?xml version="1.0" encoding="UTF-8"?>
<errors>
    <error field="subscription.base" symbol="already_subscribed">You already have a subscription to this plan.</error>
</errors>
```